### ISSUE TITLE
Table Partitioning in Postgres

### ISSUE THREAD

**Issue by ansarizafar** [prisma/prisma2#1708]
Is it possible to create a partitioned table with schema language?

**Comment by pantharshit00** [comment]
Right now we don't support this feature so I am marking this as a feature request.

**Comment by ansarizafar** [comment]
Declarative table partitioning is very important for query performance in production. Its a must have feature and Prisma should include this feature in Version 1.0

**Comment by Gmulti** [comment]
Is this functionality planned for the near future? :)

**Comment by Reiiya** [comment]
I am also very interested in this feature - if it is within year's plans?

**Comment by AkifumiSato** [comment]
I also need this feature.
Is there any way to achieve a partition at this time?

**Comment by janpio** [comment]
Most probably this is transparent for Prisma Client when it already exists in the database. Can one of you share the minimal SQL required to set up a table with partitioning? Then we can test this a bit and see if that is true.

Note: You can definitely _not_ represent the partitioning information of a table in Prisma Schema yet for it to be created by Prisma Migrate.

**Comment by AkifumiSato** [comment]
@janpio Does this fit what you are looking for?

```sql
CREATE TABLE measurement (
    city_id         int not null,
    logdate         date not null,
    peaktemp        int,
    unitsales       int
) PARTITION BY RANGE (logdate);
CREATE TABLE measurement_y2019m01 PARTITION OF measurement
    FOR VALUES FROM ('2019-01-01') TO ('2019-02-01');

CREATE TABLE measurement_y2019m02 PARTITION OF measurement
    FOR VALUES FROM ('2019-02-01') TO ('2019-03-01');

CREATE TABLE measurement_y2019m03 PARTITION OF measurement
    FOR VALUES FROM ('2019-03-01') TO ('2019-04-01');
```

**Comment by janpio** [comment]
Yes, now everyone has the context what it is we are talking about above @AkifumiSato - thanks!

So Prisma schema would need to represent 1. the `PARTITION BY RANGE (logdate);` thing for the main table, and then the 3 additional `CREATE TABLE ... PARTITION of ...` in some way so the partitions itself can also be represented. As they logically belong to the main table, this would probably make sense to all bind to the `model` in Prisma Schema (vs. the legacy representation as their own tables in the raw SQL).

The partition (table) naming is arbitrary?

**Comment by AkifumiSato** [comment]
Sorry for replying so late.

> The partition (table) naming is arbitrary?

yes, the partition naming is arbitrary.
It seems that postgresql also has `PARTITION BY LIST` and` PARTITION BY HASH`.

**Comment by janpio** [comment]
This also exists in MySQL: https://dev.mysql.com/doc/refman/8.0/en/partitioning.html

**Comment by Suniron** [comment]
Oh! I really need this 😍

**Comment by MrLoh** [comment]
It appears that you can simply setup your partitions with sql and then declar each partition as a separate table in your prisma schema, as far as I can tell.

**Comment by janpio** [comment]
Issue with SQL examples to reproduce the current failure of Prisma was here: https://github.com/prisma/prisma/issues/4667

**Comment by janpio** [comment]
Question to everyone posting here and interested in partition support in Prisma:

How would you expect Prisma to represent partitioned tables and their partitions?

In https://github.com/prisma/prisma/issues/10870 and https://github.com/prisma/prisma/issues/17348 we have two users that created separate models for each of the partitions (additionally to the main table) - a sensible approach with what Prisma produced on Introspection/`db pull` on partitioned tables in older version. Unfortunately, that badly fails in other parts of Prima's tooling (`migrate dev` specifically).

**Comment by MrLoh** [comment]
Yeah one model per partition and one for the main table would make sense to me.

**Comment by janpio** [comment]
We improved Introspection for partitioned tables in 4.10.0. This should allow everyone to manually modify the migration with `migrate dev --create-only` to put the correct SQL - and then partitioned tables should just work!

**Comment by pclokcer** [comment]
or migrate command may not say db reset ?

**Comment by janpio** [comment]
We'll create public documentation for this soon. Right now we suggest you only model the main table with a `model`, and create the table and partition tables manually. Then interact with the partitions via raw queries if you have to. Is that adequate?

**Comment by DewZ89** [comment]
Hi @revero-doug 
Could you post some code snippets?

**Comment by DewZ89** [comment]
Indeed. 
Thanks

**Comment by ZhaoKunLong** [comment]
> Question to everyone posting here and interested in partition support in Prisma:
> 
> How would you expect Prisma to represent partitioned tables and their partitions?
> 
> (In #10870 and #17348 we have two users that created separate models for each of the partitions (additionally to the main table) - a sensible approach with what Prisma produced on Introspection/`db pull` on partitioned tables in older version. Unfortunately, that badly fails in other parts of Prima's tooling (`migrate dev` specifically).)

Hello @janpio  This is my case, 
- I'm using the AWS RDS and the PostgreSQL is 12.14.
- So there was a log table `log_his`  have (id, createdAt) columns. You know It will become bigger and bigger over time.
- So I need to set `log_his` as the main table and partitioned, and use [pg_partman](https://github.com/pgpartman/pg_partman/tree/master)  to manage the partition tables.
- After I've done these on db.  I prefer to do the query using Prisma. 
-  I express that in `scheme.prisma` has one model for the main table, And has a decorator like `@Partitioned`, Then in the service can still just use `prisma.LogHis.findMany()` that would be nice!

**Comment by RyKilleen** [comment]
@ZhaoKunLong's expectations meet mine! (very similar use case.)

**Comment by mohammedpatla** [comment]
Yea, I am looking for a use case where I can auto-create partitions on postgress. The business case for us is we have a time series table, and we want to bulk the data on a monthly data set so that we can drop data after a certain period of months pass. So, with partitions, we have the ability to drop old months in Bulk without hitting the delete data option on the table.

**Comment by DanhezCode** [comment]
please add support for table partitions this is a very used feature

**Comment by snspinn** [comment]
Reading through the postgres docs on this, seems like this is straight forward enough for table writes (write to the parent table and postgres takes care of the rest) but to be able to do reads you'd need some sort of dynamic SQL. This due to fact that you have to be able to reference changing table names, especially if you want to take advantage of indexing and joins.

I don't think Prisma has support for dynamic table selection yet https://github.com/prisma/prisma/discussions/9534 and this doesn't really seem to fit into the syntax of an ORM well.

I haven't seen these points brought up in this thread yet. They seem underappreciated hurdles to get this working, or am I missing something?

**Comment by DanhezCode** [comment]
You're right that it's probably a big effort to provide support for dynamic table selection, but I disagree that this doesn't fit into the syntax of an orm because I think it can be a great extra functionality for people who handle large amounts of data and need partition-based scaling for better performance. If you decide to support this functionality, the community can surely help you, including me if you need it.

**Comment by snspinn** [comment]
> You're right that it's probably a big effort to provide support for dynamic table selection, but I disagree that this doesn't fit into the syntax of an orm because I think it can be a great extra functionality for people who handle large amounts of data and need partition-based scaling for better performance. If you decide to support this functionality, the community can surely help you, including me if you need it.

Not a developer on this, just wanted to acknowledge how much work this looks like. I retract my statement on it not fitting into the syntax: It does seem like it could work. Similar to the suggestion made here https://github.com/prisma/prisma/issues/1122#issuecomment-1662072367 regarding multi-schema support, perhaps one could expand ZhaoKunLong's suggestion to be something like
``` js
prisma.model.findMany({ partition: noneDefaultParitition })  // partition default (when blank) = parent table
```

**Comment by gustaveWPM** [comment]
Hello. I love the DX of Prisma and would be very glad to see it built-in too!

**Comment by ghyath5** [comment]
4 years ago 😁 
And still this request opened, it's such an important feature I wish prisma includes it

**Comment by gustaveWPM** [comment]
> 4 years ago 😁 And still this request opened, it's such an important feature I wish prisma includes it

Still, Prisma has come a long way since then.
PostgreSQL is a rabbit hole, and Prisma implements a lot of cursed features that nobody has any idea of to properly support what it offers in its way of standardisation.

That doesn't mean I don't hope with all my heart that we'll be able to have a 'Magic' partitioning thanks to Prisma in the year to come. It would clearly be one more killer feature.

**Comment by nscott** [comment]
Just wanted to cross-reference this thread with https://github.com/prisma/prisma/issues/7944#issuecomment-2520471168: even if partitioning is supported, that still leaves FDW as another hurdle past that. There's a minimal reproduction for both partitioning and FDW usage in the linked comment.

**Comment by CharlieDigital** [comment]
> Reading through the postgres docs on this, seems like this is straight forward enough for table writes (write to the parent table and postgres takes care of the rest) but to be able to do reads you'd need some sort of dynamic SQL. This due to fact that you have to be able to reference changing table names, especially if you want to take advantage of indexing and joins.
> 
> I don't think Prisma has support for dynamic table selection yet [#9534](https://github.com/prisma/prisma/discussions/9534) and this doesn't really seem to fit into the syntax of an ORM well.
> 
> I haven't seen these points brought up in this thread yet. They seem underappreciated hurdles to get this working, or am I missing something?
> 
> _**Note:**_ _I'm referring more to how this would be used in queries, rather than generating of migrations from the schema file._

Why would it need dynamic SQL?  The docs indicate that this is transparent and the planner will run the correct query based on partition: https://www.postgresql.org/docs/current/ddl-partitioning.html#DDL-PARTITION-PRUNING

**Comment by ukpagrace** [comment]
This is still open, is the feature added now?

**Comment by cesarvspr** [comment]
Up

**Comment by v-g-ustinov** [comment]
Up

**Comment by tuncatunc** [comment]
I asked this to grok and this is the answer.
Basically, You can manually do that using migrations, It won't effect your prisma quieries.
https://grok.com/share/bGVnYWN5_5149e433-a954-4e09-9850-e7016e36d31a

**Comment by gitnlsn** [comment]
I want this feature too. I have prisma in all projects i maintain. Partitioning would facilitate things a lot.

**Comment by domantas-jurkus-dcd** [comment]
up

**Comment by Kila14** [comment]
Up

**Comment by bijink** [comment]
I also need this feature.

**Comment by holinov** [comment]
Up

**Comment by netojoseaugusto** [comment]
Up

**Comment by iamthewalkingkode** [comment]
Up
